### PR TITLE
clear relative callback of Buffer if MessageFilter is destroyed

### LINKED ISSUE #703: rviz2 randomly crashes while running with nav2 stack
### Basic info
RViz2 from frequently crashes during the use while running with Nav2 stack. The problem appears in random places even if all Nav2-related stuff was disabled (see minimal configuration file in reproduction section below). Crashes were [noted](https://github.com/ros-planning/navigation2/issues/2367) on Crystal and still appearing on latest Rolling.

### More info
Crashes appears irregularly in different places. All types of messages and backtraces were detected and sorted-out by their probability of occurrence:
1. Invalid Argument
```
terminate called after throwing an instance of 'std::system_error'
  what():  Invalid argument
```
2. Segfault
```
Thread 16 "rviz2" received signal SIGSEGV, Segmentation fault.
```
3. Assertion Robust
```
rviz2: ../nptl/pthread_mutex_lock.c:428: __pthread_mutex_lock_full: Assertion `e != ESRCH || !robust' failed.
--Type <RET> for more, q to quit, c to continue without paging--
```
4. Deadlock: RViz2 hangs during the use
5. Assertion Prio
```
[rviz2-4] rviz2: tpp.c:82: __pthread_tpp_change_priority: Assertion `new_prio == -1 || (new_prio >= fifo_min_prio && new_prio <= fifo_max_prio)' failed.
[ERROR] [rviz2-4]: process has died [pid 174985, exit code -6, cmd '/home/leha/ros2_rolling/install/rviz2/lib/rviz2/rviz2 -d /home/leha/nav2_ws/install/nav2_bringup/share/nav2_bringup/rviz/nav2_default_view.rviz --ros-args'].
```

For the types No.1-4 GDB backtraces were collected and attached in the archive: [backtraces.zip](https://github.com/ros2/rviz/files/6618073/backtraces.zip)
For type No.5 backtrace was not obtained (it was detected once). All of the problems above were detected by using the same scenario. Thus maybe (but not proven) all issue types have the same root cause.

### Steps to reproduce

Despite the problem is irregular and there is no 100% algorithm how to reproduce it, here are the steps allowing with more or less probability to get the problem:
1. Replace `navigation2/nav2_bringup/bringup/rviz/nav2_default_view.rviz` file with minimal repro-configuraiton attached to current issue: [nav2_default_view.rviz.zip](https://github.com/ros2/rviz/files/6618076/nav2_default_view.rviz.zip)
2. Run Nav2 simulation as usual:
```
ros2 launch nav2_bringup tb3_simulation_launch.py
```
3. In appeared RViz screen set robot pose by "2D Pose Estimate"
4. enable-disable many times the "Controller" checkbox as shown in the picture below:
reproduction
5. If problem won't appear within ~30-40 seconds, close RViz and repeat starting from step No.2.

### Environment
* Ubuntu 20.04 x64
* ROS2 Distro: Rolling built from sources (`ros2.repos` from Jun 1st).
* Nav2 main and Nav2 dependencies were [built from sources](https://navigation.ros.org/build_instructions/index.html#manually-build-main). Nav2 commit: https://github.com/ros-planning/navigation2/commit/b63b11f8a638faf389d6b99911d95899abe1580b
* RViz, Qt, OGRE, OpenGl version as printed by rviz:
RViz commit: 94f6dd6958b2927a307893ff174f182dcc7afd0e
```
$ ros2 run rviz2 rviz2 -l
...
[INFO] [1623165832.507667729] [rviz2]: *-*-* OGRE Initialising
[INFO] [1623165832.507680536] [rviz2]: *-*-* Version 1.12.1 (Rhagorthua)
...
[INFO] [1623166434.886964038] [rviz2]: OpenGl version: 4.6 (GLSL 4.6)
```

Qt version is not printed by rviz:
```
$ qmake --version
Using Qt version 5.12.8 in /usr/lib/x86_64-linux-gnu
```
* If source build, which git commit?
The problem is long-standing detected since Crystal. Accused commit was not found.

### PR DESCRIPTION
**Pull request by iuhilnehc-ynos**
to fix https://github.com/ros2/rviz/issues/703

Signed-off-by: Chen Lihui <lihui.chen@sony.com>

### CODE REVIEW

File: tf2_ros/include/tf2_ros/async_buffer_interface.h [hunk 1]
```diff
@@ -39,11 +40,44 @@
...
 #include <string>
+#include <utility>
 
 namespace tf2_ros
 {
 
-using TransformStampedFuture = std::shared_future<geometry_msgs::msg::TransformStamped>;
+class TransformStampedFuture : public std::shared_future<geometry_msgs::msg::TransformStamped>
+{
+  typedef std::shared_future<geometry_msgs::msg::TransformStamped> _Base_type;
```
Comment (aprotyas) [inline tf2_ros/include/tf2_ros/async_buffer_interface.h]: Nit. Just to maintain consistency with other `using` directives in this patch.
```suggestion
  using _Base_type = std::shared_future<geometry_msgs::msg::TransformStamped>;
```

File: tf2_ros/include/tf2_ros/message_filter.h [hunk 2]
```diff
@@ -423,12 +432,17 @@ class MessageFilter : public MessageFilterBase, public message_filters::SimpleFi
...
       const auto & target_frame = std::get<2>(param);
-      buffer_.waitForTransform(
+      tf2_ros::TransformStampedFuture future = buffer_.waitForTransform(
         target_frame,
         frame_id,
         stamp,
         buffer_timeout_,
         std::bind(&MessageFilter::transformReadyCallback, this, std::placeholders::_1, handle));
+
+      if (0 != future.getHandle() || 0xffffffffffffffffULL != future.getHandle()) {
```
Comment (aprotyas) [inline tf2_ros/include/tf2_ros/message_filter.h:445]: Nit. Does `0xffffffffffffffffULL` represent the maximum `uint64_t` value? If so, wouldn't `std::numeric_limits<uint64_t>::max()` be better?

File: tf2_ros/include/tf2_ros/async_buffer_interface.h [hunk 3]
```diff
@@ -39,11 +40,44 @@
...
+
+public:
+  /// Constructor
+  explicit TransformStampedFuture(_Base_type && future) noexcept
+  : _Base_type(std::move(future)) {}
+
+  /// Copy constructor
+  TransformStampedFuture(const TransformStampedFuture & ts_future) noexcept
+  : _Base_type(ts_future),
```
Comment (aprotyas) [inline tf2_ros/include/tf2_ros/async_buffer_interface.h]: Just curious: what does this type cast achieve for all the constructors in this class?

**Review by aprotyas — COMMENTED** [review]
Some nits inline.

File: tf2_ros/include/tf2_ros/async_buffer_interface.h (hunk 3 above)
Comment (clalancette) [reply tf2_ros/include/tf2_ros/async_buffer_interface.h]: I don't think it is a type cast; instead, I think it is calling the constructor of the base class with the argument.  @iuhilnehc-ynos can confirm, though.

File: tf2_ros/include/tf2_ros/buffer.h [hunk 4]
```diff
@@ -252,6 +252,14 @@ class Buffer : public BufferInterface, public AsyncBufferInterface, public tf2::
       callback);
   }
 
+  /**
+   * \brief Cancel the future to make sure the callback of requested transform is clean.
+   * \sa cancel(TransformStampedFuture& ts_future);
```
Comment (clalancette) [inline tf2_ros/include/tf2_ros/buffer.h]: Nit: In addition to the `\sa`, I think it is better to add a `\param` with documentation of the parameter.  It's a little redundant, but makes it easier on the reader of the API docs.

File: tf2_ros/include/tf2_ros/message_filter.h (hunk 2 above)
Comment (clalancette) [reply tf2_ros/include/tf2_ros/message_filter.h:445]: Yes and no.  While it *is* the maximum value, we also use it as a hard-coded constant elsewhere, like https://github.com/ros2/geometry2/blob/7a660094d0da9c463be5fea1df60d836b3349e9b/tf2_ros/src/buffer.cpp#L249 .  It is not ideal, and maybe we should instead switch to using the limit, but I think that would be a separate patch.

Comment (clalancette) [reply tf2_ros/include/tf2_ros/message_filter.h:445]: I think it is worthwhile to add a comment here that if the future is `0` or `0xffffffffffffffffULL`, then `waitForTransform` will have already called the callback.  It's certainly possible to discover this by reading the Buffer/BufferCore code, but it is not super-obvious.

**Review by clalancette — COMMENTED** [review]
Overall, this looks pretty great.  Once the review nits are fixed, I'll give another whirl on reproducing this before approving.

**Comment by iuhilnehc-ynos** [comment]
Thanks for the quick review. I have addressed your comments, please review it again.

File: tf2_ros/include/tf2_ros/async_buffer_interface.h [hunk 5]
```diff
@@ -47,7 +47,7 @@ namespace tf2_ros
 
 class TransformStampedFuture : public std::shared_future<geometry_msgs::msg::TransformStamped>
 {
-  typedef std::shared_future<geometry_msgs::msg::TransformStamped> _Base_type;
+  using _Base_type = std::shared_future<geometry_msgs::msg::TransformStamped>;
```
Comment (aprotyas) [inline tf2_ros/include/tf2_ros/async_buffer_interface.h]: The `_Base_type` identifier is reserved by the C++ standard. From [Identifiers (cppreference)](https://en.cppreference.com/w/cpp/language/identifiers):
```the identifiers that begin with an underscore followed by an uppercase letter are reserved```

How about:

```cpp
using Base_Type = std::shared_future<geometry_msgs::msg::TransformStamped>;
```

The `Base_Type` type alias is scoped within `tf2_ros::TransformStampedFuture` anyway, so there wouldn't be any identifier conflicts.

**Review by aprotyas — COMMENTED** [review]
Minor feedback about the base type identifier inline.

File: tf2_ros/include/tf2_ros/async_buffer_interface.h (hunk 5 above)
Comment (iuhilnehc-ynos) [reply tf2_ros/include/tf2_ros/async_buffer_interface.h]: Thank you. Updated in https://github.com/ros2/geometry2/pull/490/commits/d30e1878376b150177ab73152a4dfcd223284a46

Comment (iuhilnehc-ynos) [reply tf2_ros/include/tf2_ros/async_buffer_interface.h]: I need to say, it's from `/usr/include/c++/9/future`
```cpp
  /// Primary template for shared_future.
  template<typename _Res>
    class shared_future : public __basic_future<_Res>
    {
      typedef __basic_future<_Res> _Base_type;

    public:
```

**Review by aprotyas — APPROVED** [review]
Thanks for iterating - LGTM with green CI.
@iuhilnehc-ynos can CI be run on this separately or should CI be run on both this PR and https://github.com/ros2/rviz/pull/809 together?

**Comment by iuhilnehc-ynos** [comment]
@aprotyas 

`tf2_ros` doesn't depend on `rviz` packages.
this PR can run separately if CI uses the `--packages-up-to tf2_ros`.

https://github.com/ros2/rviz/pull/809 must be run together with this PR.

**Comment by aprotyas** [comment]
Running CI against downstream users of `tf2_ros` as well just to be sure.

Repos file: https://gist.githubusercontent.com/aprotyas/e6ad6c1585b8951c8ce4ab35ccf86e1d/raw/688a29ff4e6acf2d90c92d05b9e74c3c73a78864/ros2.repos
Build args: `--packages-above-and-dependencies tf2_ros`
Test args: `--packages-above tf2_ros`
Job: https://ci.ros2.org/job/ci_launcher/9574/

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15855/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10532/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13520/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16150/)

**Comment by aprotyas** [comment]
@iuhilnehc-ynos I think we need to run CI against the changes in https://github.com/ros2/rviz/pull/809 as well since `tf2_ros::AsyncBufferInterface::cancel` is a pure virtual method that needs to be overridden. I'll wait for that to be in a testable state before re-running CI.

---

Edit: My bad, of course CI will fail if I don't build and test only up to `tf2_ros`. I'll re-run CI only up to `tf2_ros` as a first pass, but I think we should wait to build and test packages downstream of `tf2_ros` before merging.

CI:
Repos file: https://gist.githubusercontent.com/aprotyas/e6ad6c1585b8951c8ce4ab35ccf86e1d/raw/688a29ff4e6acf2d90c92d05b9e74c3c73a78864/ros2.repos
Build args: `--packages-up-to tf2_ros`
Test args: `--packages-select tf2_ros`
Job: https://ci.ros2.org/job/ci_launcher/9575/

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15856/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10533/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13521/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16151/)

**Comment by aaronchongth** [comment]
Hello @iuhilnehc-ynos! I am currently still not able to reproduce the error from the original issue (checking and unchecking the message filter in rviz, still trying at the moment), however after looking at the implementation, I am concerned that this fix will only apply to message filters as `Buffer::cancel` still to be called during the clean-up.

The test below can be added into `tf_ros2/test/test_buffer.cpp` to test for use-after-free scenarios, specifically after calling `Buffer::waitForTransform` but having the buffer going out of scope without calling `cancel`,

```
TEST(test_buffer, wait_for_transform_buffer_destructed_race)
{
  rclcpp::Clock::SharedPtr clock = std::make_shared<rclcpp::Clock>(RCL_SYSTEM_TIME);
  bool callback_timeout = false;
  std::unique_ptr<tf2_ros::TransformStampedFuture> future = nullptr;

  {
    tf2_ros::Buffer buffer(clock);
    // Silence error about dedicated thread's being necessary
    buffer.setUsingDedicatedThread(true);
    auto mock_create_timer = std::make_shared<MockCreateTimer>();
    buffer.setCreateTimerInterface(mock_create_timer);

    rclcpp::Time rclcpp_time = clock->now();
    tf2::TimePoint tf2_time(std::chrono::nanoseconds(rclcpp_time.nanoseconds()));

    future = std::make_unique<tf2_ros::TransformStampedFuture>(
      buffer.waitForTransform(
        "foo",
        "bar",
        tf2_time, tf2::durationFromSec(1.0),
        [&callback_timeout](const tf2_ros::TransformStampedFuture & future)
        {
          try {
            // We don't expect this throw, even though a timeout will occur
            future.get();
          } catch (...) {
            callback_timeout = true;
          }
        }));
  }

  auto status = future->wait_for(std::chrono::seconds(1));
  EXPECT_EQ(status, std::future_status::timeout);
}
```

I've created a PR a while ago, https://github.com/ros2/geometry2/pull/480, that resolves the use-after-free issue on the `Buffer` level, but haven't had the time to revisit the problem with the overall `nav2` and `rviz` test. Could you perhaps test if my branch fixes the issues for your scenario?

**Comment by iuhilnehc-ynos** [comment]
> Could you perhaps test if my branch fixes the issues for your scenario?

Hi @aaronchongth, I'd like to do the test, but I am afraid that https://github.com/ros2/geometry2/pull/480 does not fix the https://github.com/ros2/rviz/issues/703#issue-915219371.

```sh
chenlh geometry2 $ git remote -v
aaronchongth	https://github.com/aaronchongth/geometry2.git (fetch)
aaronchongth	https://github.com/aaronchongth/geometry2.git (push)
clh	git@github.com:iuhilnehc-ynos/geometry2.git (fetch)
clh	git@github.com:iuhilnehc-ynos/geometry2.git (push)
origin	https://github.com/ros2/geometry2.git (fetch)
origin	https://github.com/ros2/geometry2.git (push)

[00:01:26 KINGSTON] /home/chenlh/Projects/ROS2/ros2-master/src/ros2/geometry2 (aaron-buffer-shared-data) 
chenlh geometry2 $ git branch -vv
* aaron-buffer-shared-data     3e76762 [aaronchongth/aaron/buffer-shared-data] Added additinoal test for resetting a unique ptr of a buffer
  ros2                         9ff7dad [origin/ros2] tf2: Enable common linter tests (#469)
  topic-cancel-future-callback d30e187 [clh/topic-cancel-future-callback] not to use underscore that is reserved by the C++ standard
```

Rebuild ros2 workspaces(I put ros2-master and navigation2 in a workspace) with `--packages-above-and-dependencies tf2 tf2_ros rviz_common rviz_default_plugins`, rviz2 is blocked or crashed by using the following test method.
(Test five times with the method: run with `ros2 launch nav2_bringup tb3_simulation_launch.py`, and quickly uncheck and check the `Controller` group multiple times(e.g 20), finally set `2D Pose Estimate` for the robot. You can continue to uncheck and check the `Controller` group if it works well.)

use https://github.com/ros2/geometry2/pull/490 and https://github.com/ros2/rviz/pull/809 and rebuild them again, rviz2 seems to work well by the above test method.

(NOTE: I test the whole procedure twice.)

**Comment by aaronchongth** [comment]
hello @iuhilnehc-ynos! Thanks for the advice and testing, I was finally able to reproduce the crashes by running them on my laptop, and you are right, my PR does not fix the bug. With your solution the problem doesn't occur in all the 20 runs I tried, thanks for the instructions! All good on my end now.

Some information about how I was testing:

* Commented out the simulation launch in `nav2_bringup/launch/tb3_simulation.launch.py`, so I don't need to keep launching the sim,
```python
    # Add any conditioned actions
    # ld.add_action(start_gazebo_server_cmd)
    # ld.add_action(start_gazebo_client_cmd)
    # ld.add_action(start_gazebo_spawner_cmd)
```

* launch sim once
```bash
ros2 launch turtlebot3_gazebo turtlebot3_world.launch.py
```

* launch, test, Ctrl+C, repeat
```bash
ros2 launch nav2_bringup tb3_simulation_launch.py
```

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-2-tsc-meeting-january-20th-2022/23986/1

**Review by clalancette — APPROVED** [review]
This looks fantastic to me; it's just the kind of solution that I couldn't figure out on my own.  I'm going to go ahead and review the rviz2 fix as well, and then run CI on the lot.

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/16178/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10806/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16520/)

**Comment by clalancette** [comment]
The warnings on aarch64 aren't because of this PR; see https://ci.ros2.org/job/ci_linux-aarch64/10812 which does not have this change, but has the same warning.

Since everything else is happy here, I'm going to go ahead and merge.

**Comment by ijnek** [comment]
While investigating another issue, I came across this piece of code:

```
// If handle of future is 0 or 0xffffffffffffffffULL, waitForTransform have already called
// the callback.
if (0 != future.getHandle() || 0xffffffffffffffffULL != future.getHandle()) {.....}
```

This evaluates to true for any value. Looking at the conversations that have happened, perhaps it could be one of the following:

```
if (0 != future.getHandle() && 0xffffffffffffffffULL != future.getHandle())
if (!(0 == future.getHandle() || 0xffffffffffffffffULL != future.getHandle()))
```

Unfortunately I don't have enough knowledge about what this does to test out changes... @iuhilnehc-ynos or @clalancette could you confirm this is an issue?

**Comment by iuhilnehc-ynos** [comment]
@ijnek 

Thank you. It is an issue.

```
1. ! (0 == handle || 0xffffffffffffffffULL == handle)

2. (0 != handle && 0xffffffffffffffffULL != handle)
                //\  it should be &&
```

**Pull request merged**